utils/litemap/src/store/vec_impl.rs

```diff
@@ -104,38 +104,170 @@ impl<K, V> StoreMut<K, V> for Vec<(K, V)> {
+        self.extend(iter);
+        // `sorted_len` is the length of the sorted run before extension
+        // window slice `w` is guaranteed to have a length of 2.
+        #[allow(clippy::indexing_slicing)]
+        {
+            // Count new elements that are sorted and non-duplicated.
+            // Starting from the end of the existing sorted run, if any.
+            sorted_len += self[sorted_len.saturating_sub(1)..]
+                .windows(2)
+                .take_while(|w| w[0].0 < w[1].0)
+                .count()
+                + (sorted_len == 0) as usize;
+        }
+        // If everything was in order, we're done
+        if sorted_len >= self.len() {
+            return;
+        }
+
+        // Use stable sort to keep relative order of duplicates.
+        self.sort_by(|a, b| a.0.cmp(&b.0));
+        // Deduplicate by keeping the last element of the run in the first slice.
+        let (dedup, _merged_dup) = partition_dedup_by(self, |a, b| a.0 == b.0);
+        sorted_len = dedup.len();
+        self.truncate(sorted_len);
+    }
 }
 
-impl<K: Ord, V> StoreFromIterable<K, V> for Vec<(K, V)> {
-    fn lm_sort_from_iter<I: IntoIterator<Item = (K, V)>>(iter: I) -> Self {
-        let iter = iter.into_iter();
-        let mut container = match iter.size_hint() {
-            (_, Some(upper)) => Self::with_capacity(upper),
-            (lower, None) => Self::with_capacity(lower),
-        };
-
-        for (key, value) in iter {
-            if let Some(last) = container.lm_last() {
-                if last.0 >= &key {
-                    match container.lm_binary_search_by(|k| k.cmp(&key)) {
-                        #[allow(clippy::unwrap_used)] // Index came from binary_search
-                        Ok(found) => {
-                            let _ =
-                                core::mem::replace(container.lm_get_mut(found).unwrap().1, value);
-                        }
-                        Err(ins) => {
-                            container.insert(ins, (key, value));
-                        }
-                    }
-                } else {
-                    container.push((key, value))
-                }
+/// Moves all but the _last_ of consecutive elements to the end of the slice satisfying
+/// a given equality relation.
+///
+/// Returns two slices. The first contains no consecutive repeated elements.
+/// The second contains all the duplicates in no specified order.
+///
+/// This is based on std::slice::partition_dedup_by (currently unstable) but retains the
+/// _last_ element of the duplicate run in the first slice (instead of first).
+#[inline]
+fn partition_dedup_by<T, F>(v: &mut [T], mut same_bucket: F) -> (&mut [T], &mut [T])
+where
+    F: FnMut(&mut T, &mut T) -> bool,
+{
+    // Although we have a mutable reference to `self`, we cannot make
+    // *arbitrary* changes. The `same_bucket` calls could panic, so we
+    // must ensure that the slice is in a valid state at all times.
```

Comment:
Thought: I assume these docs are based on comments elsewhere, but we can guarantee that `same_bucket` won't panic.

Reply:
Good point